internal/querycoordv2/checkers/segment_checker.go

```diff
@@ -0,0 +1,242 @@
+		redundancies = append(redundancies, segments...)
+	}
+	tasks = c.createSegmentReduceTasks(ctx, redundancies, replica.GetID(), querypb.DataScope_Streaming)
+	ret = append(ret, tasks...)
+
+	return ret
+}
+
+func (c *SegmentChecker) getSegmentsDist(replica *meta.Replica) []*meta.Segment {
+	ret := make([]*meta.Segment, 0)
+	for _, node := range replica.Nodes.Collect() {
+		ret = append(ret, c.dist.SegmentDistManager.GetByCollectionAndNode(replica.CollectionID, node)...)
+	}
+	return ret
+}
+
+func diffSegments(targets []*datapb.SegmentInfo, dists []*meta.Segment) (lacks []*datapb.SegmentInfo, redundancies []*meta.Segment) {
+	distMap := make(map[int64]struct{})
+	targetMap := make(map[int64]struct{})
+	for _, s := range targets {
+		targetMap[s.GetID()] = struct{}{}
+	}
+	for _, s := range dists {
+		distMap[s.GetID()] = struct{}{}
+		if _, ok := targetMap[s.GetID()]; !ok {
+			redundancies = append(redundancies, s)
+		}
+	}
+	for _, s := range targets {
+		if _, ok := distMap[s.GetID()]; !ok {
+			lacks = append(lacks, s)
+		}
+	}
+	return
+}
+
+func findRepeatedSegments(dists []*meta.Segment) []*meta.Segment {
+	ret := make([]*meta.Segment, 0)
+	versions := make(map[int64]*meta.Segment)
+	for _, s := range dists {
+		maxVer, ok := versions[s.GetID()]
+		if !ok {
+			versions[s.GetID()] = s
+			continue
+		}
+		if maxVer.Version <= s.Version {
+			ret = append(ret, maxVer)
+			versions[s.GetID()] = s
+		} else {
+			ret = append(ret, s)
+		}
+	}
+	return ret
+}
+
+func (c *SegmentChecker) filterExistedOnLeader(replica *meta.Replica, segments []*meta.Segment) []*meta.Segment {
+	filtered := make([]*meta.Segment, 0, len(segments))
+	for _, s := range segments {
+		leaderID, ok := c.dist.ChannelDistManager.GetShardLeader(replica, s.GetInsertChannel())
+		if !ok {
+			continue
+		}
+		onLeader := false
+		leaderViews := c.dist.LeaderViewManager.GetLeaderView(leaderID)
+		for _, view := range leaderViews {
+			node, ok := view.Segments[s.GetID()]
+			if ok && node == s.Node {
+				onLeader = true
+				break
+			}
+		}
+		if onLeader {
+			// if this segment is serving on leader, do not remove it for search available
+			continue
+		}
+		filtered = append(filtered, s)
+	}
+	return filtered
+}
+
+func (c *SegmentChecker) findNeedReleasedGrowingSegments(replica *meta.Replica) map[int64][]int64 {
+	// log := log.With(
+	// 	zap.Int64("collectionID", replica.GetCollectionID()),
+	// 	zap.Int64("replicaID", replica.GetID()),
+	// )
+
+	ret := make(map[int64][]int64, 0) // leaderID -> segment ids
+	segments := c.targetMgr.GetSegmentsByCollection(replica.GetCollectionID())
+	leaders := c.dist.ChannelDistManager.GetShardLeadersByReplica(replica)
+	leaderViews := make(map[string]*meta.LeaderView) // shard -> leaderview
+	for _, s := range segments {
+		lview, ok := leaderViews[s.GetInsertChannel()]
+		if !ok {
+			leaderID, ok := leaders[s.GetInsertChannel()]
+			if !ok {
+				// log.Warn("leader not found", zap.String("channel", s.GetInsertChannel()))
```

Comment:
leaderViews will never be updated if leaderID does not exist, that is an exception and should handle it.

Reply:
> never be updated if leaderID does not exi

It will be fixed in channel checker.